File: leetCode/easy/14-longestCommonPrefix.ts
import { assertEquals } from "https://deno.land/std@0.182.0/testing/asserts.ts";
import {
  describe,
  it,
  beforeEach,
  afterEach,
} from "https://deno.land/std@0.182.0/testing/bdd.ts";

function longestCommonPrefix(strs: string[]): string {
  if (strs.length === 0) return "";
  if (strs.length === 1) return strs[0];

  let charIdx = 0;
  let commonSet = new Set<string>();
  const commonChars: string[] = [];

  // find the min length of arr to use as the loop limit
  const minLength = Math.min(...strs.map((str) => str.length));

  // amount of times to keep looping through strs
  let arrLoop = 0;

  while (arrLoop < minLength) {
    // loop through strs and add each char to the set
    strs.forEach((str) => {
      const char = str[charIdx];
      commonSet.add(char);
    });

    // if set length === 1 there is a common prefix and is added to the commonChars array
    if (commonSet.size === 1) {
      commonChars.push(Array.from(commonSet.values())[0]);
    } else commonChars.push(" ");

    // reset set and increment charIdx and arrLoop
    commonSet = new Set<string>();
    charIdx += 1;
    arrLoop += 1;
  }

  // if the first char is a space, then no common prefix
  return commonChars[0] === " "
    ? ""
    : commonChars.join("").trim().split(" ")[0];
}

export { longestCommonPrefix };

// console.log(longestCommonPrefix(["aca", "cba"]));

describe("longestCommonPrefix", () => {
  it("1. should return the longest common prefix", () => {
    const strs = ["flower", "flow", "flight"];
    assertEquals(longestCommonPrefix(strs), "fl");
  });

  it("2. should return the longest common prefix", () => {
    const strs = ["dog", "racecar", "car"];
    assertEquals(longestCommonPrefix(strs), "");
  });

  it("3. should return the longest common prefix", () => {
    const strs = ["dog", "dog", "dog"];
    assertEquals(longestCommonPrefix(strs), "dog");
  });

  it("4. should return the longest common prefix", () => {
    const strs = ["dog"];
    assertEquals(longestCommonPrefix(strs), "dog");
  });

  it("5. should return the longest common prefix", () => {
    const strs = ["cir", "car"];
    assertEquals(longestCommonPrefix(strs), "c");
  });

  it("6. should return the longest common prefix", () => {
    const strs = ["aca", "cba"];
    assertEquals(longestCommonPrefix(strs), "");
  });

  it("7. should return the longest common prefix", () => {
    const strs = ["a"];
    assertEquals(longestCommonPrefix(strs), "a");
  });

  it("8. should return the longest common prefix", () => {
    const strs = ["ab", "a"];
    assertEquals(longestCommonPrefix(strs), "a");
  });

  it("9. should return the longest common prefix", () => {
    const strs = ["babb", "caa"];
    assertEquals(longestCommonPrefix(strs), "");
  });
});

File: leetCode/easy/14-longestCommonPrefix.jl
function longestCommonPrefix(strs::Array{String,1})::String
  if length(strs) === 0
    return ""
  end

  if length(strs) === 1
    return first(strs)
  end

  # find the min length of arr to use as the loop limit
  minLength = minimum(length, strs)
  # amount of times to keep looping through strs
  arrLoop = 0
  charIdx = 1
  commonSet = Set{Char}()
  commonChars = []

  while arrLoop < minLength
    # loop through strs and add each char to the set
    foreach(str -> push!(commonSet, str[charIdx]), strs)

    # if set length === 1 there is a common prefix and is added to the commonChars array
    if length(commonSet) === 1
      char = first(values(commonSet))
      push!(commonChars, char)
    else
      push!(commonChars, ' ')
    end

    # reset set
    commonSet = Set{Char}()
    charIdx += 1
    arrLoop += 1
  end

  # if the first char is a space, then no common prefix
  if first(commonChars) === ' '
    return ""
  else
    commonCharsStr = commonChars |> join |> strip
    return first(split(commonCharsStr, ' '))
  end
end



using Test

Test.@testset "longestCommonPrefix" begin
  # leetcode tests
  @test longestCommonPrefix(["flower", "flow", "flight"]) === "fl"
  @test longestCommonPrefix(["dog", "racecar", "car"]) === ""
  @test longestCommonPrefix([""]) === ""
  @test longestCommonPrefix(["a"]) === "a"
  @test longestCommonPrefix(["a", "b"]) === ""
  @test longestCommonPrefix(["a", "a"]) === "a"
  @test longestCommonPrefix(["a", "a", "b"]) === ""
  @test longestCommonPrefix(["a", "a", "a"]) === "a"
  @test longestCommonPrefix(["cir", "car"]) === "c"
  @test longestCommonPrefix(["aa", "a"]) === "a"
  @test longestCommonPrefix(["a", "aa"]) === "a"
  @test longestCommonPrefix(["aca", "cba"]) === ""
  @test longestCommonPrefix(["babb", "caa"]) === ""
end
